### ISSUE TITLE
New rule proposal: no-template-literal-title

### ISSUE THREAD

**Issue by unlikelyzero** [mskelton/eslint-plugin-playwright#499]
## The problem

A test title is an identifier. It is what the HTML reporter groups by, what
`--grep` matches, what a flake dashboard keys on, and what a reviewer searches
for when CI fails. A template literal breaks that contract in two different
ways, so it is worth naming both.

**Interpolated titles are not stable identifiers.**

```js
test(`login as ${user.role}`, async ({ page }) => { ... })

test.describe(`${env.NAME} checkout`, () => {
  test.step(`upload ${file.name}`, async () => { ... })
})
```

The title only exists once the data does, and it changes when the data changes.
The consequences are practical:

1. **Grep stops working.** `npx playwright test --grep "login as admin"` matches
   nothing you can predict from the source. Neither does a code search for the
   title a colleague pasted from a failing run.
2. **History fragments.** A title that reads `upload invoice-8842.pdf` on one
   run and `upload invoice-9130.pdf` on the next is two different tests to any
   tool that tracks pass/fail over time. Flake detection and quarantine lists
   silently stop working.
3. **The reported name can hide the real input.** `${user.role}` prints `admin`,
   but the failure may depend on the rest of `user`. The title looks specific
   while omitting what actually mattered.

Where the data genuinely varies, the ways to say so are already visible ones:
generate distinct static titles at the loop site, put the varying value in a
`test.step` body or an annotation, or attach it with `test.info().annotations`
so it lands in the report as data rather than as the test's name.

**Static template literals are just the wrong quote.**

```js
test(`checkout completes`, async ({ page }) => { ... })
```

There is no interpolation here, so nothing is dynamic — but it is inconsistent
with every other title in the suite and it invites the interpolated form later.
This case is mechanically fixable.

## Proposed rule

`no-template-literal-title` — report a template literal used as the title of
`test`, `test.describe`, or `test.step`.

Two distinct reports:

- **Interpolated** (`` `login as ${role}` ``) — reported, no autofix. There is
  no mechanical rewrite; the author has to decide what the title should be.
- **Static** (`` `checkout completes` ``) — reported with an autofix that
  converts it to a string literal.

Applies to the title argument of `test`, `test.only`, `test.skip`, `test.fixme`,
`test.describe` (and its modifiers), and `test.step`, plus any configured
`globalAliases`.

Not reported:

- non-title arguments — a template literal in a locator, assertion, or step body
  is untouched
- `test.describe(() => {})` with no title
- titles that are string literals or concatenations of them

### Options

- `ignore: Array<'test' | 'test.describe' | 'test.step'>` — skip the listed
  callers, matching the `ignore` option already used by
  [`prefer-lowercase-title`](https://github.com/mskelton/eslint-plugin-playwright/tree/main/docs/rules/prefer-lowercase-title.md).
  Teams that deliberately generate parameterized test titles can set
  `{ ignore: ['test'] }` and still get the rule for describes and steps.

Not in the `recommended` config — parameterized titles are an established
pattern and this is a project-level policy choice.

**Comment by mskelton** [comment]
I'm not sure I agree with this. The argument of changing title based on id is only if you use a dynamically changing title. But for things like it's a totally valid way to write test titles.

```js
['admin', 'editor', 'user'].map(role => {
  test(`${role} works`)
})
```

I'm almost positive that this will pass with grep as well, since Playwright evaluates all the test names when running the test function, before they are executed.
